File: unitree_slam_example_new/global_nav/build_grid_map.py
#!/usr/bin/env python3
"""
Build a 2D occupancy grid from a .pcd/.ply point cloud map.

Usage:
  python3 build_grid_map.py \
    --input /mnt/disk1/unitree_slam_example/global_maps/global_map_20260330_155423.pcd \
    --output /mnt/disk1/unitree_slam_example/global_nav/map_data.npz
"""

from __future__ import annotations

import argparse
import logging
from pathlib import Path

import numpy as np

try:
    import open3d as o3d
except Exception as exc:  # pragma: no cover
    o3d = None
    OPEN3D_IMPORT_ERROR = exc
else:
    OPEN3D_IMPORT_ERROR = None


LOGGER = logging.getLogger("build_grid_map")


def _pcd_numpy_dtype(type_name: str, size: int) -> str:
    if type_name == "F":
        return {4: "<f4", 8: "<f8"}[size]
    if type_name == "I":
        return {1: "<i1", 2: "<i2", 4: "<i4", 8: "<i8"}[size]
    if type_name == "U":
        return {1: "<u1", 2: "<u2", 4: "<u4", 8: "<u8"}[size]
    raise ValueError(f"Unsupported PCD field type: {type_name}{size}")


def _read_pcd_header(file_obj) -> dict[str, object]:
    header: dict[str, object] = {}
    while True:
        raw_line = file_obj.readline()
        if not raw_line:
            raise ValueError("Invalid PCD file: missing DATA header")
        line = raw_line.decode("ascii", errors="strict").strip()
        if not line or line.startswith("#"):
            continue

        parts = line.split()
        key = parts[0].upper()
        values = parts[1:]
        if key == "FIELDS":
            header["fields"] = values
        elif key == "SIZE":
            header["sizes"] = [int(v) for v in values]
        elif key == "TYPE":
            header["types"] = values
        elif key == "COUNT":
            header["counts"] = [int(v) for v in values]
        elif key == "WIDTH":
            header["width"] = int(values[0])
        elif key == "HEIGHT":
            header["height"] = int(values[0])
        elif key == "POINTS":
            header["points"] = int(values[0])
        elif key == "DATA":
            header["data"] = values[0].lower()
            return header


def load_pcd_points(file_path: Path) -> np.ndarray:
    with file_path.open("rb") as file_obj:
        header = _read_pcd_header(file_obj)

        fields = header.get("fields")
        sizes = header.get("sizes")
        types = header.get("types")
        if not isinstance(fields, list) or not isinstance(sizes, list) or not isinstance(types, list):
            raise ValueError(f"Invalid PCD header in: {file_path}")
        counts = header.get("counts")
        if not isinstance(counts, list):
            counts = [1] * len(fields)
        points = int(header.get("points") or int(header.get("width", 0)) * int(header.get("height", 1)))
        data_type = str(header.get("data", ""))

        if points <= 0:
            raise ValueError(f"No points declared in PCD: {file_path}")
        if not {"x", "y", "z"}.issubset(fields):
            raise ValueError(f"PCD must contain x/y/z fields: {file_path}")

        LOGGER.info("Loading PCD file path=%s points=%s data=%s", file_path, points, data_type)
        if data_type == "binary":
            dtype_fields = []
            offsets = []
            offset = 0
            for name, size, type_name, count in zip(fields, sizes, types, counts):
                dtype = _pcd_numpy_dtype(type_name, size)
                dtype_fields.append((name, dtype) if count == 1 else (name, dtype, (count,)))
                offsets.append(offset)
                offset += size * count
            dtype = np.dtype({"names": [item[0] for item in dtype_fields],
                              "formats": [item[1:] if len(item) > 2 else item[1] for item in dtype_fields],
                              "offsets": offsets,
                              "itemsize": offset})
            data = np.fromfile(file_obj, dtype=dtype, count=points)
            if data.shape[0] != points:
                raise ValueError(f"Unexpected EOF while reading PCD: {file_path}")
            return np.column_stack([data["x"], data["y"], data["z"]]).astype(np.float64)

        if data_type == "ascii":
            data = np.loadtxt(file_obj, dtype=np.float64, max_rows=points)
            if data.ndim == 1:
                data = data.reshape(1, -1)
            field_index = {name: idx for idx, name in enumerate(fields)}
            return data[:, [field_index["x"], field_index["y"], field_index["z"]]]

        raise ValueError(f"Unsupported PCD DATA type: {data_type}")


def load_points(file_path: Path) -> np.ndarray:
    if file_path.suffix.lower() == ".pcd":
        return load_pcd_points(file_path)

    if o3d is None:
        raise RuntimeError(
            "open3d is required for non-PCD point clouds. Install with: pip install open3d"
        ) from OPEN3D_IMPORT_ERROR

    LOGGER.info("Loading point cloud with open3d path=%s", file_path)
    pcd = o3d.io.read_point_cloud(str(file_path))
    pts = np.asarray(pcd.points, dtype=np.float64)
    if pts.size == 0:
        raise ValueError(f"No points loaded from: {file_path}")
    return pts


def build_occupancy(
    points_xyz: np.ndarray,
    resolution: float,
    z_min: float,
    z_max: float,
    inflation_radius: float,
) -> tuple[np.ndarray, float, float, float]:
    # Height filter for near-ground obstacles / structure points.
    mask = (points_xyz[:, 2] >= z_min) & (points_xyz[:, 2] <= z_max)
    pts = points_xyz[mask]
    if pts.size == 0:
        raise ValueError(
            "No points left after z filtering. Relax --z-min/--z-max."
        )

    min_x = float(np.min(pts[:, 0]))
    max_x = float(np.max(pts[:, 0]))
    min_y = float(np.min(pts[:, 1]))
    max_y = float(np.max(pts[:, 1]))

    width = int(np.ceil((max_x - min_x) / resolution)) + 1
    height = int(np.ceil((max_y - min_y) / resolution)) + 1

    occ = np.zeros((height, width), dtype=np.uint8)

    gx = np.floor((pts[:, 0] - min_x) / resolution).astype(np.int32)
    gy = np.floor((pts[:, 1] - min_y) / resolution).astype(np.int32)

    valid = (gx >= 0) & (gx < width) & (gy >= 0) & (gy < height)
    gx = gx[valid]
    gy = gy[valid]
    occ[gy, gx] = 1

    # Inflate obstacles by robot radius + safety margin.
    inflate_cells = int(np.ceil(inflation_radius / resolution))
    if inflate_cells > 0:
        inflated = occ.copy()
        occupied_y, occupied_x = np.where(occ == 1)
        for cy, cx in zip(occupied_y, occupied_x):
            y0 = max(0, cy - inflate_cells)
            y1 = min(height, cy + inflate_cells + 1)
            x0 = max(0, cx - inflate_cells)
            x1 = min(width, cx + inflate_cells + 1)

            yy, xx = np.ogrid[y0:y1, x0:x1]
            circle = (yy - cy) ** 2 + (xx - cx) ** 2 <= inflate_cells**2
            inflated[y0:y1, x0:x1][circle] = 1
        occ = inflated

    return occ, min_x, min_y, resolution


def main() -> None:
    logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
    parser = argparse.ArgumentParser(description="Build 2D occupancy grid from point cloud")
    parser.add_argument("--input", required=True, help="Input .pcd/.ply map")
    parser.add_argument("--output", required=True, help="Output .npz map file")
    parser.add_argument("--resolution", type=float, default=0.10, help="Grid resolution in meters")
    parser.add_argument("--z-min", type=float, default=-0.20, help="Min z for obstacle extraction")
    parser.add_argument("--z-max", type=float, default=1.50, help="Max z for obstacle extraction")
    parser.add_argument(
        "--inflation-radius",
        type=float,
        default=0.45,
        help="Obstacle inflation radius in meters (G1 recommended: 0.45)",
    )
    args = parser.parse_args()

    input_path = Path(args.input)
    output_path = Path(args.output)
    LOGGER.info(
        "Build occupancy map input=%s output=%s resolution=%s z_min=%s z_max=%s inflation_radius=%s",
        input_path,
        output_path,
        args.resolution,
        args.z_min,
        args.z_max,
        args.inflation_radius,
    )

    points = load_points(input_path)
    occ, origin_x, origin_y, resolution = build_occupancy(
        points,
        resolution=args.resolution,
        z_min=args.z_min,
        z_max=args.z_max,
        inflation_radius=args.inflation_radius,
    )

    output_path.parent.mkdir(parents=True, exist_ok=True)
    np.savez_compressed(
        output_path,
        occupancy=occ,
        origin=np.array([origin_x, origin_y], dtype=np.float64),
        resolution=np.array([resolution], dtype=np.float64),
    )

    # Also dump a quick debug image (0=free, 255=occupied).
    try:
        from PIL import Image

        img = (occ * 255).astype(np.uint8)
        Image.fromarray(np.flipud(img), mode="L").save(output_path.with_suffix(".png"))
    except Exception:
        pass

    print("Map saved:", output_path)
    print("Grid shape (h, w):", occ.shape)
    print("Origin (x, y):", (origin_x, origin_y))
    print("Resolution:", resolution)


if __name__ == "__main__":
    main()

File: unitree_slam_example_new/global_nav/build_grid_map.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

#include <pcl/io/pcd_io.h>
#include <pcl/io/ply_io.h>
#include <pcl/point_types.h>

struct Config {
  std::string input_path;
  std::string output_path;
  double resolution = 0.10;
  double z_min = -0.20;
  double z_max = 1.50;
  double inflation_radius = 0.45;
};

struct GroundCellStats {
  int count = 0;
  double min_z = std::numeric_limits<double>::infinity();
};

static constexpr double kGroundCellSizeM = 0.20;
static constexpr int kGroundMinCellPoints = 2;
static constexpr double kGroundHistBinSizeM = 0.05;

static void PrintUsage() {
  std::cout << "Usage:\n"
            << "  build_grid_map_cpp --input <map.pcd|map.ply> --output <map.txt> "
               "[--resolution 0.1] [--z-min -0.2] [--z-max 1.5] [--inflation-radius 0.45]\n";
  std::cout << "  --z-min/--z-max are heights relative to the estimated ground plane.\n";
}

static bool ParseArgs(int argc, char** argv, Config& cfg) {
  for (int i = 1; i < argc; ++i) {
    std::string a = argv[i];
    auto need = [&](const std::string& name) -> std::string {
      if (i + 1 >= argc) {
        throw std::runtime_error("Missing value for " + name);
      }
      return argv[++i];
    };

    if (a == "--input") cfg.input_path = need(a);
    else if (a == "--output") cfg.output_path = need(a);
    else if (a == "--resolution") cfg.resolution = std::stod(need(a));
    else if (a == "--z-min") cfg.z_min = std::stod(need(a));
    else if (a == "--z-max") cfg.z_max = std::stod(need(a));
    else if (a == "--inflation-radius") cfg.inflation_radius = std::stod(need(a));
    else {
      std::cerr << "Unknown arg: " << a << "\n";
      return false;
    }
  }

  return !cfg.input_path.empty() && !cfg.output_path.empty();
}

static double Percentile(const std::vector<double>& sorted_values, double percentile) {
  if (sorted_values.empty()) {
    throw std::runtime_error("Cannot compute percentile of empty values");
  }
  const double pos = (percentile / 100.0) * (double)(sorted_values.size() - 1);
  const size_t lo = (size_t)std::floor(pos);
  const size_t hi = (size_t)std::ceil(pos);
  if (lo == hi) return sorted_values[lo];
  const double ratio = pos - (double)lo;
  return sorted_values[lo] * (1.0 - ratio) + sorted_values[hi] * ratio;
}

static double Median(std::vector<double> values) {
  if (values.empty()) {
    throw std::runtime_error("Cannot compute median of empty values");
  }
  std::sort(values.begin(), values.end());
  const size_t mid = values.size() / 2;
  if (values.size() % 2 == 1) return values[mid];
  return (values[mid - 1] + values[mid]) * 0.5;
}

static double EstimateGroundZ(const pcl::PointCloud<pcl::PointXYZ>& cloud) {
  double min_x = std::numeric_limits<double>::infinity();
  double min_y = std::numeric_limits<double>::infinity();
  size_t finite_points = 0;
  for (const auto& p : cloud.points) {
    if (!std::isfinite(p.x) || !std::isfinite(p.y) || !std::isfinite(p.z)) continue;
    min_x = std::min(min_x, (double)p.x);
    min_y = std::min(min_y, (double)p.y);
    ++finite_points;
  }
  if (finite_points == 0) {
    throw std::runtime_error("Cannot estimate ground height from empty point cloud");
  }

  std::unordered_map<int64_t, GroundCellStats> cells;
  cells.reserve(finite_points / 4 + 1);
  for (const auto& p : cloud.points) {
    if (!std::isfinite(p.x) || !std::isfinite(p.y) || !std::isfinite(p.z)) continue;
    const int64_t gx = (int64_t)std::floor(((double)p.x - min_x) / kGroundCellSizeM);
    const int64_t gy = (int64_t)std::floor(((double)p.y - min_y) / kGroundCellSizeM);
    const int64_t key = (gy << 32) ^ (gx & 0xffffffffLL);
    auto& cell = cells[key];
    ++cell.count;
    cell.min_z = std::min(cell.min_z, (double)p.z);
  }

  std::vector<double> candidates;
  candidates.reserve(cells.size());
  std::vector<double> all_cell_mins;
  all_cell_mins.reserve(cells.size());
  for (const auto& kv : cells) {
    all_cell_mins.push_back(kv.second.min_z);
    if (kv.second.count >= kGroundMinCellPoints) candidates.push_back(kv.second.min_z);
  }
  if (candidates.empty()) {
    candidates = all_cell_mins;
    std::cout << "INFO Ground estimation fallback: using all lower-envelope cells\n";
  }
  if (candidates.empty()) {
    throw std::runtime_error("No lower-envelope cells available for ground estimation");
  }

  std::sort(candidates.begin(), candidates.end());
  const double lo = Percentile(candidates, 1.0);
  const double hi = Percentile(candidates, 99.0);
  std::vector<double> trimmed;
  trimmed.reserve(candidates.size());
  for (double z : candidates) {
    if (z >= lo && z <= hi) trimmed.push_back(z);
  }
  if (trimmed.empty()) trimmed = candidates;

  const double min_z = *std::min_element(trimmed.begin(), trimmed.end());
  const double max_z = *std::max_element(trimmed.begin(), trimmed.end());
  const double bin_start = std::floor(min_z / kGroundHistBinSizeM) * kGroundHistBinSizeM;
  const int bin_count = std::max(1, (int)std::ceil((max_z - bin_start) / kGroundHistBinSizeM) + 1);
  std::vector<int> hist((size_t)bin_count, 0);
  for (double z : trimmed) {
    int bin = (int)std::floor((z - bin_start) / kGroundHistBinSizeM);
    bin = std::max(0, std::min(bin_count - 1, bin));
    ++hist[(size_t)bin];
  }

  const int best_bin = (int)std::distance(hist.begin(), std::max_element(hist.begin(), hist.end()));
  const double mode_lo = bin_start + (double)best_bin * kGroundHistBinSizeM;
  const double mode_hi = mode_lo + kGroundHistBinSizeM;
  std::vector<double> mode_values;
  for (double z : trimmed) {
    if (z >= mode_lo && z < mode_hi) mode_values.push_back(z);
  }

  const double ground_z = Median(mode_values.empty() ? trimmed : mode_values);
  std::cout << "INFO Estimated ground_z=" << ground_z << " from lower envelope cells=" << all_cell_mins.size()
            << " candidates=" << candidates.size() << " mode_bin=[" << mode_lo << ", " << mode_hi << ")\n";
  return ground_z;
}

int main(int argc, char** argv) {
  Config cfg;
  try {
    if (!ParseArgs(argc, argv, cfg)) {
      PrintUsage();
      return 1;
    }
  } catch (const std::exception& e) {
    std::cerr << e.what() << "\n";
    PrintUsage();
    return 1;
  }
  if (cfg.z_min > cfg.z_max) {
    std::cerr << "Invalid height range: --z-min must be <= --z-max\n";
    return 1;
  }

  pcl::PointCloud<pcl::PointXYZ> cloud;
  const auto ext = std::filesystem::path(cfg.input_path).extension().string();
  int ret = -1;
  if (ext == ".pcd") {
    ret = pcl::io::loadPCDFile(cfg.input_path, cloud);
  } else if (ext == ".ply") {
    ret = pcl::io::loadPLYFile(cfg.input_path, cloud);
  } else {
    std::cerr << "Unsupported input extension: " << ext << "\n";
    return 1;
  }

  if (ret < 0 || cloud.empty()) {
    std::cerr << "Failed to load point cloud: " << cfg.input_path << "\n";
    return 1;
  }

  std::cout << "INFO Build occupancy map input=" << cfg.input_path << ", output=" << cfg.output_path
            << ", resolution=" << cfg.resolution << ", relative_z_min=" << cfg.z_min
            << ", relative_z_max=" << cfg.z_max << ", inflation_radius=" << cfg.inflation_radius << "\n";

  double ground_z = 0.0;
  try {
    ground_z = EstimateGroundZ(cloud);
  } catch (const std::exception& e) {
    std::cerr << "Failed to estimate ground height for " << cfg.input_path << ": " << e.what() << "\n";
    return 1;
  }
  const double abs_z_min = ground_z + cfg.z_min;
  const double abs_z_max = ground_z + cfg.z_max;
  std::cout << "INFO Applying ground-relative height filter ground_z=" << ground_z
            << ", relative_z=[" << cfg.z_min << ", " << cfg.z_max << "]"
            << ", absolute_z=[" << abs_z_min << ", " << abs_z_max << "]\n";

  std::vector<pcl::PointXYZ> pts;
  pts.reserve(cloud.size());
  for (const auto& p : cloud.points) {
    if (p.z >= abs_z_min && p.z <= abs_z_max) pts.push_back(p);
  }

  if (pts.empty()) {
    std::cerr << "No points after z filter, relax --z-min/--z-max\n";
    return 1;
  }

  double min_x = pts[0].x, max_x = pts[0].x;
  double min_y = pts[0].y, max_y = pts[0].y;
  for (const auto& p : pts) {
    min_x = std::min(min_x, (double)p.x);
    max_x = std::max(max_x, (double)p.x);
    min_y = std::min(min_y, (double)p.y);
    max_y = std::max(max_y, (double)p.y);
  }

  const int width = (int)std::ceil((max_x - min_x) / cfg.resolution) + 1;
  const int height = (int)std::ceil((max_y - min_y) / cfg.resolution) + 1;
  std::vector<uint8_t> occ((size_t)width * height, 0);

  auto idx = [width](int x, int y) { return (size_t)y * width + x; };

  for (const auto& p : pts) {
    int gx = (int)std::floor((p.x - min_x) / cfg.resolution);
    int gy = (int)std::floor((p.y - min_y) / cfg.resolution);
    if (gx >= 0 && gx < width && gy >= 0 && gy < height) {
      occ[idx(gx, gy)] = 1;
    }
  }

  const int inflate_cells = (int)std::ceil(cfg.inflation_radius / cfg.resolution);
  if (inflate_cells > 0) {
    auto inflated = occ;
    for (int y = 0; y < height; ++y) {
      for (int x = 0; x < width; ++x) {
        if (occ[idx(x, y)] == 0) continue;
        int y0 = std::max(0, y - inflate_cells);
        int y1 = std::min(height - 1, y + inflate_cells);
        int x0 = std::max(0, x - inflate_cells);
        int x1 = std::min(width - 1, x + inflate_cells);
        for (int yy = y0; yy <= y1; ++yy) {
          for (int xx = x0; xx <= x1; ++xx) {
            int dx = xx - x;
            int dy = yy - y;
            if (dx * dx + dy * dy <= inflate_cells * inflate_cells) {
              inflated[idx(xx, yy)] = 1;
            }
          }
        }
      }
    }
    occ.swap(inflated);
  }

  std::filesystem::create_directories(std::filesystem::path(cfg.output_path).parent_path());
  std::ofstream ofs(cfg.output_path);
  if (!ofs) {
    std::cerr << "Cannot write: " << cfg.output_path << "\n";
    return 1;
  }

  // Custom map text format:
  // line1: width height
  // line2: origin_x origin_y resolution
  // line3+: each row as 0/1 chars
  ofs << width << " " << height << "\n";
  ofs << min_x << " " << min_y << " " << cfg.resolution << "\n";
  for (int y = 0; y < height; ++y) {
    for (int x = 0; x < width; ++x) ofs << (occ[idx(x, y)] ? '1' : '0');
    ofs << "\n";
  }

  std::cout << "Saved map: " << cfg.output_path << "\n";
  std::cout << "Grid size: " << width << " x " << height << "\n";
  std::cout << "Origin: (" << min_x << ", " << min_y << "), resolution=" << cfg.resolution << "\n";
  std::cout << "Estimated ground z: " << ground_z << "\n";
  std::cout << "Absolute z filter: (" << abs_z_min << ", " << abs_z_max << ")\n";

  return 0;
}
